`HfQuantizer` class for quantization-related stuff in `modeling_utils.py`

poedator opened the pull request:
### What does this PR do?
Refactoring `modeling_utils.py` to move all quantization-related logic into new `HFQuantizer` class

### Reasons and benefits:
- easier to understand how quantization works during loading.
- easier to add new quantization methods like SPQR.
- much easier to implement 4-bit serialization (once BnB supports it)
- it was a rainy day outside )

### Things to be done:
- extend `HFQuantizer` functionality to cover `_load_pretrained_model()` and `_load_state_dict_into_model()`
- extend `HFQuantizer` functionality in BnB to cover `set_module_quantized_tensor_to_device()` from `integrations/bitsandbytes.py`
- review multiple TODOs left in the comments.
- undo temporary changes in the tests
- check for backward compatibility issues
- eliminate repeats and redundancies, if any
- consider fully absorbing `integrations/bitsandbytes.py` into new class
- move new classes / functions into proper project file. possibly separate folder.

### Current state:
- reworked code up to `_load_state_dict_into_model()`
- all tests in BnB and GPTQ still pass with `RUN_SLOW=1`.

summoning  @SunMarc and @younesbelkada to comment on the idea and current state.

younesbelkada commented:
Thanks a lot for your huge work! I really like the idea! Let me know once this is ready for a first pass, I'll have a deep look

poedator commented:
Hi, @younesbelkada!
By now I reworked `_load_pretrained_model()` and `_load_state_dict_into_model()` functions. 
`Integations.bitsandbytes.py::set_module_quantized_tensor_to_device()` is replicated in new code and no longer needed.
I like that by having separate `HFQuantizer` classes there is no need to have multiple ifs in the code.

The code works fine and passes all RUN_SLOW quantization tests: gptq, 4/8 bit bnb.
It is now ready for the first detailed review - please take a look.

It may be further optimized, but I tried to preserve some resemblance with the earlier code.
Three are quite a few TODOs left - some of them require your guidance.
Please recommend ways to simplify the `HFQuantizer` class, to make it easier to implement new quantizers. 
For instance, there are few small functions to validate and adjust environment. Could they be consolidated? Also, shall we separate fresh quantization code from loading pre-quantized?

younesbelkada commented on src/transformers/utils/quantization_config.py:
```diff
@@ -48,7 +49,7 @@ class QuantizationConfigMixin:
 
     @classmethod
-    def from_dict(cls, config_dict, return_unused_kwargs=False, **kwargs):
+    def from_dict(cls, config_dict, return_unused_kwargs=False, config_origin="config", **kwargs):
```
Can you elaborate more on the utility of `config_origin` argument? Do you think that we can find a way to overcome the usage of this argument?

younesbelkada replied:
Ah yes that sounds great! Thanks for the suggestion, if you can make it with `quantization_status` usong an enum I think it will make it much more readble indeed!

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+
+
+class QuantizationConfigParser:
+    def __init__(self):
```
could be nice to add some docstring here

poedator replied:
done

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+    @abstractmethod
+    def method(self):
+        """placeholder to method from QuantizationMethod"""
+        pass
```
```suggestion
        ...
```
You can just set `...` instead of `pass` I think since it is an abstract method

poedator replied:
done

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+        self.config_origin = config_origin
+        self.modules_to_not_convert = []
+
+    def validate_environment(self, *args, **kwargs):
+        """checking env after instantiation"""
+        pass
```
you can also make that abstract

poedator replied:
done

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+    def adjust_max_memory(self, max_memory):
+        return max_memory
+
+    def validate_device_map(self, device_map):
+        pass
```
you can also make that abstract

poedator replied:
done

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+        super().process_model_after_weight_loading(model)
+
+
+class BnbEightBitHFQuantizer(BnbHFQuantizer):
```
Do you think it would make sense to just declare `Bnb8bitHFQuantizer` and `Bnb4bitHFQuantizer` since only 2 methods are using `super()` - what do you think?

poedator replied:
@younesbelkada, Besides these 2 methods with `super()` there are 8 other shared methods in `BnbHFQuantizer`. They correspond places in the original code with condition `load_in_8bit or load_in_4bit`.
Actually there is some common code in `create_quantized_param()` and `process_model_before_weight_loading()` to be moved into `BnbHFQuantizer` class.

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+        return unexpected_keys, missing_keys
+
+
+class BnbFourBitHFQuantizer(BnbHFQuantizer):
```
```suggestion
class Bnb4BitHFQuantizer(BnbHFQuantizer):
```

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+        super().process_model_after_weight_loading(model)
+
+
+class BnbEightBitHFQuantizer(BnbHFQuantizer):
```
```suggestion
class Bnb8BitHFQuantizer(BnbHFQuantizer):
```

younesbelkada reviewed (commented):
Thanks a lot for your huge work ! I really like the refactor and the cleanup, this makes things much cleaner in `from_pretrained`. I left few open questions before some final reviews, let me know what do you think!

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -38,7 +38,7 @@
 from .generation import GenerationConfig, GenerationMixin
 from .integrations import PeftAdapterMixin, deepspeed_config, is_deepspeed_zero3_enabled
-from .pytorch_utils import (  # noqa: F401
+from .pytorch_utils import (  # noqa: F401  # pylint: disable=unused-import
```
This came from my linter on VSCode. Will remove it.

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+        self.quantization_config_from_args = quantization_config_from_args
+        load_in_4bit = kwargs.pop("load_in_4bit", False)
+        load_in_8bit = kwargs.pop("load_in_8bit", False)
+        # TODO: consider deprecation notice for quant kwargs outside quantization_config
```
added a deprecation notice

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+            quantization_method = self.quantization_method_from_args
+            config_origin = "args" if self.quantization_config_from_args is not None else None
+        except AttributeError:
+            raise AttributeError("`quantization_config_from_args` must be parsed before parsing model config")
```
done

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,715 @@
+
+            quantization_config_kwargs = {
+                k: v for k, v in kwargs.items() if k in inspect.signature(BitsAndBytesConfig).parameters
+            }
```
Done. makes sense 100%

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -3230,39 +2983,24 @@ def from_pretrained(
-                    )
-                del device_map_without_lm_head
+            if quantizer is not None:
+                quantizer.validate_device_map
```
I had the same thought but hesitated to combine these two methods because because between 
`quantizer.update_device_map()` and `quantizer. validate_device_map()`:
- changes are made to `device_map`, by `infer_auto_device_map()` 
- `modules_to_not_convert` are updated during quantizer.process_model_before_weight_loading()
Let me revisit this in the future, after some testing...

poedator commented:
@younesbelkada , I addressed most of your comments and left couple open. Take a look.

please, suggest how to address the errors in CircleCI. Most are not quite clear to me

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,741 @@
+        self.quantization_status = quantization_status
+        self.modules_to_not_convert = []
+
+    def validate_environment(self, *args, **kwargs):
```
```suggestion
    @abstractmethod
    def validate_environment(self, *args, **kwargs):
```

poedator replied:
Could we keep it as is?  Some `HFQuantizer` subclasses may choose not to care about the env and not redefine this method.

younesbelkada replied:
Hmm we could always let users define a "dummy" `validate_environment` right?

poedator replied:
agreed. I see benefits in the need for the developers to double-check their environment requirements.

younesbelkada replied:
Makes sense!

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -3230,32 +2983,16 @@ def from_pretrained(
-                    )
-                del device_map_without_lm_head
+            if quantizer is not None:
+                quantizer.validate_device_map
```
Then this should be : 

```suggestion
                quantizer.validate_device_map(device_map)
```

No?

younesbelkada reviewed (commented):
Thanks a lot @poedator for your great work, my main comment is only: https://github.com/huggingface/transformers/pull/26610/files#r1364571404 it would be great if you can use an enum instead! 
After that I will battle test your PR locally and let you know!

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -3320,13 +3057,10 @@ def from_pretrained(
                 keep_in_fp32_modules=keep_in_fp32_modules,
             )
 
-        model.is_loaded_in_4bit = load_in_4bit
-        model.is_loaded_in_8bit = load_in_8bit
```
it is taken care of in `HFQuantizer.process_model_before_weight_loading()`. You may want to deprecate it in the future and replace with ref to `model.config.quantization_config`

poedator commented:
@younesbelkada ,
I changed to enum, added typing. 
let's try these battle tests.

younesbelkada commented:
Thanks so much @poedator ! Will try that out by today and let you know

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26610). All of your documentation changes will be reflected on that endpoint.

younesbelkada commented:
hi @poedator , sorry did not had time to have a deep look today, will do it tomorrow first thing in the morning! Thanks again for your patience! 🙏

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,787 @@
+        logger.info("Activating quantized loading for this model using {self.__class__}")
+        return model
+
+    def process_model_after_weight_loading(self, model: PreTrainedModel, **kwargs):
+        """setting model attributes and/or converting model AFTER weights loading"""
+        model.eval()
```
I would remove this and make it abstract as well, `model.eval()` is always call after `from_pretrained`
```suggestion
    @abstractmethod
    def process_model_after_weight_loading(self, model: PreTrainedModel, **kwargs):
        ...
```

younesbelkada commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,787 @@
+    method = QuantizationMethod.GPTQ
+
+    def __init__(self, quantization_config: QuantizationConfig, **kwargs):
+        super().__init__(quantization_config, **kwargs)
```
shouldn't we call `validate_environment` inside init as well? What do you think?

poedator replied:
The present choice of arguments to `validate_environment()` looks awkward to me: torch_dtype, from_tf, from_flax. 
It was driven by validation requirements of present quantizers. Passing them via quantization_config_parser.get_quantizer() would be even more awkward. 

I would rather keep the functions separate, but pass the whole context to the validation fn. This would make life easier for the future quantizers.

poedator replied:
@younesbelkada, How would you like the idea of passing the whole context of `from_pretrained()` to `HFQuantizer` methods? It is not very pythonic, but this way we'd be able to combine `validate_environment(), set_torch_dtype(), update_device_map()` into one method.

younesbelkada replied:
> I would rather keep the functions separate, but pass the whole context to the validation fn. This would make life easier for the future quantizers.

Yes OK got it, makes sense, let's keep it as it is i would say. I personally really like the fact that everything is separated and makes things very clear, validate dtype, device_map, etc. Let's wait for other maintainers to see their thoughts

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2584,69 +2568,9 @@ def from_pretrained(
-                )
+        # processing quantization args before model_kwargs instantiated
+        quantization_config_parser = QuantizationConfigParser()
+        kwargs = quantization_config_parser.parse_config_from_args(quantization_config, **kwargs)
```
just a heads up that this modifies the kwargs, but it only does

```python
        load_in_4bit = kwargs.pop("load_in_4bit", False)
        load_in_8bit = kwargs.pop("load_in_8bit", False)
```

So it is safe, maybe just adding a comment here would help

poedator replied:
added comment on this to quantizers.py::57

younesbelkada replied:
OK thanks!

younesbelkada reviewed (commented):
Looks really nice! The refactor is amazing and makes things much easier to follow and understand @poedator I have battle tested evertyhing and looks great on my end! 
I re-ran BNB tests and ran some quantization training with PEFT and all tests seem to pass so far! 
I left 3 questions - can you also help fix the merge conflicts? I will also let @ArthurZucker & @SunMarc have a look! Before merging I'll re-run the tests a last time before merging
Would be great if we can merge this PR before: https://github.com/huggingface/transformers/pull/27045 as well 🙏

poedator commented:
> ...can you also help fix the merge conflicts?

I fixed the merge conflicts, however it cancelled some of the changes from https://github.com/huggingface/transformers/pull/27043 by @patrickvonplaten.  
As I see it, keeping changes to the config makes no effect, since `process_model_after_weight_loading()` only returns the model. Please recheck, whether and how to reapply changes from #27043. Perhaps you have some test to check if the problem solved by 27043 reappears.

poedator commented:
Ideas for before and after merge:
- move quantization code to `quantize` folder - it will contain quantization_config.py, quantizers.py and files for future quant classes
- should there be more documentation/blog post on how to build a new quantizer? 
- for future serialization I have strong feeling that it is best that all respective quantizers support `to/from state_dict()` methods.
- the checks for `is_loaded_in_4bit` should now be replaced to checks based on model.config.quantization_config.

younesbelkada commented:
Thanks for all your work and for addressing all comments!

> I fixed the merge conflicts, however it cancelled some of the changes from https://github.com/huggingface/transformers/pull/27043 by @patrickvonplaten.
As I see it, keeping changes to the config makes no effect, since process_model_after_weight_loading() only returns the model. Please recheck, whether and how to reapply changes from https://github.com/huggingface/transformers/pull/27043. Perhaps you have some test to check if the problem solved by 27043 reappears.

Sure yes I will check that out myself and confirm you here

> should there be more documentation/blog post on how to build a new quantizer? The AWQ https://github.com/huggingface/transformers/pull/27045 may be a good basis for this.

Definitely 💯 ! we can always do it after this PR being merged as a follow up, so no worries I would say

poedator commented:
@younesbelkada,
I see that #27045 is merged now. I guess it is time to update this PR again -  I can try it over the weekend. Please confirm that this makes sense, or is there any other way forward?

younesbelkada commented:
hi @poedator , thanks for getting back , yes what you have said makes sense we still want to merge this PR. On my end the tests I performed were good, I can try again once you merge with main! 🙏

poedator commented:
@younesbelkada , @SunMarc , @ArthurZucker, 
The PR is rebased to `main` as of Nov 3,  including the AWQ changes. It is ready for review - please take a look.

major issues:
- seemingly unrelated test fails in `examples_torch` 
- some tests with multiple GPUs fail - need more time to debug
- the number of methods in HFQuantizer bothers me. Some could be combined if we pass context from above. See separate note
- there is still a number of TODOs left - need hints on how to approach them.
- get rid of `is_loaded_in_4[8]bit` use in `modeling_utils.py`. (see `.save_pretrained()` and `.num_parameters()`) - use quant_config refs.
- can we drop `is_loaded_in_4[8]bit` completely? or replace with @property with warning?

poedator commented:
Here are the calls to `HFQuantizer` from `modeling_utils.py`. Would be great to reduce their number. 
So far I only see it possible with contextlib. See `HFQuantizer.get_locals_from_above()` as prototype.


after initialization:
-	`quantizer.validate_environment(torch_dtype=torch_dtype, from_tf=from_tf, from_flax=from_flax)  # odd args selection`
-	`quantizer.set_torch_dtype(torch_dtype)`
-	`quantizer.update_device_map(device_map)`
-	# the calls in this group could be combined with `__init__()` if we pass context from .from_pretrained()

before weight loading:
-	quantizer.process_model_before_weight_loading(model, device_map, torch_dtype, keep_in_fp32_modules)
-	quantizer.get_special_dtypes_update(model, torch_dtype)
-	quantizer.adjust_target_dtype(target_dtype)
-	quantizer.adjust_max_memory(max_memory)
-	quantizer.validate_device_map(device_map)
the calls in this group could be combined into `process_model_before_weight_loading` if we pass context from .from_pretrained()

in `_load_pretrained_model()`:
-	`quantizer.check_quantized_param(model, param_value=value, param_name=key, state_dict={})`
-	`quantizer.update_mismatched_keys(unexpected_keys, missing_keys)`

in `_load_state_dict_into_meta_model()`:
-	`quantizer.create_quantized_param` 

after weight loading:
-	`quantizer.process_model_after_weight_loading(model)`

poedator commented:
@ArthurZucker , @SunMarc , could you comment on this PR please.
cc @younesbelkada

ArthurZucker commented:
Yeah sorry, it's a big PR and not super time sensitive so I've been pushing it back 😓 will try to do this week

SunMarc commented:
cc @Titus-von-Koeller for visibility

SunMarc commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+        # TODO: consider moving parts common with 8bits to super()
+        from ..integrations import get_keys_to_not_convert, replace_with_bnb_linear
+
+        llm_int8_skip_modules = (
+            self.quantization_config.llm_int8_skip_modules
+        )  # Despite the name, this still occurs in 4-bit q-configs!
```
Yeah, we need to deprecate this arg

poedator replied:
will remove the comment for now. Will assume that deprecation will happen in some other PR.

SunMarc reviewed (commented):
Thanks for the refactoring @poedator! This greatly improve the readability of the `from_pretrained`. Left a couple of comments. I'm for the design you proposed but I will let the core maintainer comment on that.

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2675,76 +2657,11 @@ def from_pretrained(
-                )
+        # processing quantization args before model_kwargs instantiated
+        quantization_config_parser = QuantizationConfigParser()
+        kwargs = quantization_config_parser.parse_config_from_args(quantization_config, **kwargs)
```
Great point, let's maybe address this in a follow up PR!

poedator replied:
I added deprecation warning in quantizers.py::parse_config_from_args()

poedator committed:
squashed earlier commits for easier rebase

poedator committed:
4bit save enabled @quantizers

poedator committed:
rm rebase leftovers

poedator committed:
TMP gptq test use exllama

poedator committed:
fix AwqConfigTest::test_wrong_backend for A100

poedator committed:
quantizers AWQ fixes

poedator committed:
_load_pretrained_model low_cpu_mem_usage branch

poedator committed:
quantizers style

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -737,20 +733,17 @@ def _load_state_dict_into_meta_model(
-            state_dict_index = offload_weight(param, param_name, state_dict_folder, state_dict_index)
-        elif not is_quantized:
-            # For backward compatibility with older versions of `accelerate`
+            state_dict_index = offload_weight(param, param_name, model, state_dict_folder, state_dict_index)
```
yes, looks like a typo from rebase. fixed in the fresh rebase

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -91,7 +89,8 @@
 )
-from .utils.quantization_config import AwqConfig, BitsAndBytesConfig, GPTQConfig, QuantizationMethod
+from .utils.quantization_config import QuantizationMethod
+from .utils.quantizers import QuantizationConfigParser  # TODO remove BnbHFQuantizer
```
agreed

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -2775,136 +2692,18 @@ def from_pretrained(
+            quantizer.validate_environment(torch_dtype=torch_dtype, from_tf=from_tf, from_flax=from_flax)
 
-            # Force-set to `True` for more mem efficiency
             if low_cpu_mem_usage is None:
                 low_cpu_mem_usage = True
```
added logger.warning                
`logger.warning("'low_cpu_mem_usage' was None, now set to True since model is quantized.")`

poedator committed:
remove require_low_cpu_mem_usage attr

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -3226,7 +3026,7 @@ def from_pretrained(
             logger.info("Detected DeepSpeed ZeRO-3: activating zero.init() for this model")
             init_contexts = [deepspeed.zero.Init(config_dict_or_path=deepspeed_config())] + init_contexts
-        elif load_in_8bit or load_in_4bit or low_cpu_mem_usage:
+        elif low_cpu_mem_usage or hasattr(quantizer, "require_low_cpu_mem_usage"):
```
removed `require_low_cpu_mem_usage` attribute,  assume that it is `True` for all quantized models

poedator committed:
rm dtype arg from process_model_before_weight_loading

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -3246,94 +3046,10 @@ def from_pretrained(
-                "All non-linear modules will be loaded in full precision."
-                " If you want to load the other modules in other precision, please specify a `torch_dtype` attribute."
+        if quantizer is not None:
+            quantizer.process_model_before_weight_loading(
+                model=model, device_map=device_map, torch_dtype=torch_dtype, keep_in_fp32_modules=keep_in_fp32_modules
```
Agreed that we can remove torch_dtype entirely from here.

poedator committed:
rm config_origin from Q-config

poedator commented on src/transformers/utils/quantization_config.py:
```diff
@@ -222,6 +228,16 @@ def __init__(
+    @classmethod
+    def from_dict(cls, config_dict, return_unused_kwargs=False, config_origin="config", **kwargs):
+        if (config_origin == "args") and any(set(kwargs).intersection(inspect.signature(cls).parameters)):
+            raise ValueError(
+                "You can't pass `load_in_8bit` or any other `BitsAndBytesConfig` argument as a kwarg when passing "
+                "`quantization_config` argument at the same time."
+            )
```
agree. done.

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+
+    def parse_config_from_args(self, quantization_config_from_args=None, **kwargs) -> Dict[str, Any]:
+        """
+        Parse quantization config from the arguments passed to `from_pretrained`. Sets the `quantization_config` and
+        `quantization_method` attributes based on args.
+            quantization_config_from_args: The `quantization_config` passed to `from_pretrained`. **kwargs: Additional
+            kwargs passed to `from_pretrained`, like `load_in_8bit`.
+        Returns:
+            Additional kwargs with quantization kwargs popped.
```
done

poedator committed:
rm inspect from q_config

poedator committed:
fixed docstrings in QuantizationConfigParser

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+        load_in_4bit = kwargs.pop("load_in_4bit", False)
+        load_in_8bit = kwargs.pop("load_in_8bit", False)
+        if load_in_4bit or load_in_8bit:
+            warnings.warn(
+                "The `load_in_4bit` and `load_in_8bit` arguments are deprecated and will be removed in the future versions. "
+                "Please, pass a `quantization_config` dict or instance instead.",
+                FutureWarning,
+            )
```
done

poedator committed:
logger.warning fix

poedator committed:
mv is_loaded_in_4(8)bit to BnbHFQuantizer

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+    """
+
+    use_keep_in_fp32_modules = True
+    require_low_cpu_mem_usage = True
```
removed `require_low_cpu_mem_usage`, kept `use_keep_in_fp32_modules` (still used)

poedator committed:
is_accelerate_available error msg fix in quantizer

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+
+        if not (is_accelerate_available() and is_bitsandbytes_available()):
+            raise ImportError(
+                "Using `load_in_8bit=True` requires Accelerate: `pip install accelerate` and the latest version of"
+                " bitsandbytes `pip install -i https://test.pypi.org/simple/ bitsandbytes` or"
+                " pip install bitsandbytes` "
+            )
```
fixed. 
Interesting, why `https://test.pypi.org/simple/` was here originally? - looks like a leftover from some earlier unstable version. By now releases are pushed to the regular `https://test.pypi.org/simple`. Changed that.

SunMarc replied:
cc @younesbelkada

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+        )
+        # TODO: consider bringing replace_with_bnb_linear() code from ..integrations/bitsandbyter.py to here
+
+        model.config.quantization_config = self.quantization_config
+        model.is_8bit_serializable = self.is_model_serializeable()
```
agree. will think of best way to address...

poedator replied:
`is_8bit_serializable` is still used in `.save_pretrained()` 

Moved these 2 lines to `process_model_after_weight_loading`.

combined `.process_model_before_weight_loading()` methods for 4 and 8 bits.

poedator committed:
split is_model_trainable in bnb quantizer class

poedator committed:
rm llm_int8_skip_modules as separate var in Q

poedator committed:
Q rm todo

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+        super().__init__(quantization_config, **kwargs)
+        from optimum.gptq import GPTQQuantizer
+
+        if isinstance(quantization_config, dict):
+            quantization_config = GPTQConfig.from_dict(quantization_config)
```
not sure about this. For now, from_pretrained allows passing quantization config as dict.

SunMarc replied:
Yes, indeed. However, you are already converting the **quantization_config** to an object here. : 

 ```python 
  self.quantization_config = config_class.from_dict(
            config_dict=dict(self.quantization_config)
        )  # dict to class instance

config.quantization_config = self.quantization_config
quantizer = quantizer_class(
    quantization_config=self.quantization_config, quantization_status=quantization_status
)
return quantizer
```

poedator replied:
got it now, forgot about the conversion above. removed these lines. Thanks!

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+        super().process_model_after_weight_loading(model, **kwargs)
+
+        # Note: optimum.gptq.GPTQQuantizer.post_init_model()
+        # creates model.config.quantize_config
+        # which splits functionality with model.quantization_config
+        # (also inserted by optimum.gptq.GPTQQuantizer as dict)
```
Is there anything specific that needs to be done here?

SunMarc replied:
Just to remove the comment as it is unnecessary

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+                "You cannot pass an `AwqConfig` when loading a model as you can only use AWQ models"
+                " for inference. To quantize transformers models with AWQ algorithm, please refer to our"
+                " quantization docs: https://huggingface.co/docs/transformers/main_classes/quantization "
+            )
```
removed this ValueError, added code for fusion. pls check

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+
+        return kwargs
+
+    def get_quantizer(self, config: PretrainedConfig) -> "HFQuantizer":
```
it was a way to create a type hint with a class, which is defined lower in the text.
see [PEP-484](https://peps.python.org/pep-0484/#forward-references)
But with `from __future__ import annotations` it works without quotes [PEP-563](https://peps.python.org/pep-0563/).

poedator committed:
fwd ref to HFQuantizer in type hint

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+    def __init__(self, quantization_config: QuantizationConfig, **kwargs):
+        super().__init__(quantization_config, **kwargs)
+
+        if isinstance(quantization_config, dict):
+            self.quantization_config = BitsAndBytesConfig.from_dict(quantization_config, return_unused_kwargs=False)
```
not sure about this. For now, from_pretrained allows passing quantization config as dict.

SunMarc replied:
see comments above

poedator replied:
removed

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+        elif isinstance(quantization_config, BitsAndBytesConfig):
+            self.quantization_config = quantization_config
+        else:
+            raise ValueError(
+                f"Invalid type for `quantization_config`: {type(quantization_config)}. Should be a `dict` or a"
+                " `BitsAndBytesConfig` instance."
+            )
```
- not sure about this. For now, from_pretrained allows passing quantization config as dict.

SunMarc replied:
See comments above.

poedator replied:
I removed all tests for `dict` type from all classes, also removed the validation of config type from bnb case entirely.  I assume that wrong config type would not pass thru `QuantizationConfigParser` anyways.
Now `self.quantization_config = quantization_config` is in the parent class `HFQuantizer`

SunMarc commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+            )
+
+        self.quantization_method = None
+        if self.quantization_config is not None:
+            # BNB used as default method if Q config present but has no quant_method  # TODO: consider warning / raising error.
+            self.quantization_method = getattr(
+                self.quantization_config, "quant_method", QuantizationMethod.BITS_AND_BYTES
+            )
```
Passing a dict as an arg for `quantization_config` is only properly handled for bitsandbytes. Now that we have multiple quantization scheme, we need to make sure that the dict contain a `quantization_method` value.

poedator replied:
addressed this in `QuantizationConfigParser.validate_quantization_config()`. For now the code issues a warning.

SunMarc replied:
I will have a look but in my opinion, we should probably raise an error.

poedator replied:
changed to raising error. see `QuantizationConfigParser.validate_quantization_config()`

poedator replied:
reverted to warning in case `load_in_8/4bit` key is True. 
reason: `"ybelkada/bloom-1b7-8bit"` in the 8bit tests has quantization config in dict form.

poedator committed:
rm note re optimum.gptq.GPTQQuantizer

poedator committed:
quantization_config in __init__ simplified

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -3714,12 +3384,19 @@ def _fix_key(key):
-                        set_module_quantized_tensor_to_device(
-                            model, key, "cpu", torch.empty(*param.size(), dtype=target_dtype)
-                        )
+                        raise NotImplementedError("Loading quantized parameters outside base model is not supported.")
+                        # TODO: consider whether to support this (e.g. quantized model head).
+                        # may need to create such model with quantized head for testing.
```
This is one of the places not covered by tests, so initially I made these branches throw error.
Now to be consistent with the code behavior before,  I used `quantizer.create_quantized_param()`. Is it correct to leave "cpu" here (as it was before), or shall the device be determined from the "device_map"?
`set_module_quantized_tensor_to_device()` was not a valid choice since I got rid of it in `_load_state_dict_into_meta_model()` already.

poedator committed:
replaced NonImplemented with  create_quantized_param

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -3879,22 +3556,31 @@ def _find_mismatched_keys(
                                     )
+                                    # What is FSDP? The above line is not reachable in tests.
+                                    # TODO: explore when this is needed and cover by tests.
+                                    # Was: `set_module_quantized_tensor_to_device(model, key, "cpu", torch.empty(*param.size(), dtype=target_dtype))`
```
answered above )
fixed with `quantizer.create_quantized_param()`

poedator committed:
rm load_in_4/8_bit deprecation warning

poedator committed:
QuantizationConfigParser refactoring

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+                        " (e.g. disable_exllama, use_cuda_fp16, max_input_length) will be overwritten with the one you passed to `from_pretrained`. "
+                        "The rest will be ignored."
+                    )
+                    quantization_status = QuantizationStatus.PREQUANTIZED
```
simplified a bit. pls check

is it correct that you want get_loading_attributes always have priority over config.quantization_config?

SunMarc replied:
Yes, `get_loading_attributes` is used to modify `config.quantization_config` so that we can for example change the default kernel (gptq) or use fusion (awq).

poedator replied:
get_loading_attributes take precedence in this PR too, with warning.
see `QuantizationConfigParser.handle_configs_collision()`

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,879 @@
+
+        return kwargs
+
+    def get_quantizer(self, config: PretrainedConfig) -> "HFQuantizer":
```
simplified, pls.check
quantization status  is now set in one place, effectively matching with config origin.

younesbelkada commented:
Hi @poedator ! Happy new year and thanks for putting so much efforts on this PR? Would you be happy helping us finalising the PR ? Otherwise happy to take it over as well as we want to have this refactor ASAP on transformers!

poedator commented:
Hi, @younesbelkada ,
I am ready to finalize this together with you and your team.  For now I see the following open issues:
- completing response to comments of @SunMarc 
- rebase to resolve conflicts (hope that this is trivial this time)
- updating `save_prequantized` (optional)
- another round of tests and reviews ?

younesbelkada commented:
Hi @poedator thanks for getting back ! I think it would be great if you could first resolve the merge conflicts, then complete the final comments from Marc, after that we'll do a final review and run the tests if that sounds good to you

poedator committed:
awq-related minor changes

poedator committed:
awq-related changes

poedator committed:
awq config.modules_to_not_convert

poedator committed:
raise error if no q-method in q-config in args

poedator committed:
minor cleanup

poedator committed:
awq quantizer docstring

poedator committed:
combine common parts in bnb process_model_before_weight_loading

poedator committed:
revert test_gptq

poedator committed:
.process_model_ cleanup

poedator committed:
restore dict config warning

poedator commented:
The PR code has been rebased to the latest main (as of 2023.01.07).  It is pretty much ready for the next round of reviews @younesbelkada @SunMarc @ArthurZucker @amyeroberts 

Most of the comments of Marc are resolved, a few are addressed but left in the open form for Marc to revisit.
One open item is # TODO: consider removing used param_parts from state_dict before return. I want to do some more tests before committing it.

I left a few TODOs in the quantizers.py for the maintainers to consider choices - please comment on them or delete.

There are further optimisation opportunities in this refactoring - I did not try to pursue them actively. 
Examples:
- refactor the corresponding code in `accelerate:: ... ::set_module_tensor_to_device()`
- refactor save_pretrained()
- replace `is_loaded_in_4bit` and `is_4bit_serializable` with references to config
- integrations/bitsandbytes can be integrated into quantizers.py (set_module_quantized_tensor_to_device is not used in the main code)
- quantizers.py may be sliced into smaller files (one per method)

The Long tests pass with 2 exceptions
- FAILED tests/quantization/gptq/test_gptq.py::GPTQTest::test_change_loading_attributes - AssertionError: True != False
- FAILED tests/quantization/gptq/test_gptq.py::GPTQTest::test_serialization - ValueError: Found modules on cpu/disk. ...
Not sure yet what caused them, maybe they are specific to my setup. Will update here if I find the causes.

ArthurZucker commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,869 @@
+            if quant_method is None:
+                if quantization_config.get("load_in_8bit", False) or quantization_config.get("load_in_4bit", False):
+                    logger.warning(
+                        "The model's quantization_config has no `quant_method` attribute, but has `load_in_8bit` or `load_in_4bit` parameter set. "
+                        "It is recommended to pass quantization config as `BitsAndBytesConfig`. "
```
let's also deprecate these (not twice but either here or in the modeling utils)

poedator replied:
@younesbelkada suggested to do the same, but in the follow-up PR. 
I had this deprecation notice before, but removed in [this commit](https://github.com/huggingface/transformers/pull/26610/commits/bb5c35a089b6a395ae3502f2dbb97c3e8272eedd)

ArthurZucker replied:
Alright no worries !

ArthurZucker commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,869 @@
+                    raise ValueError(
+                        "The model's quantization config from the arguments has no `quant_method` attribute."
+                    )
+
+            if quant_method == QuantizationMethod.BITS_AND_BYTES:
+                config_class = BitsAndBytesConfig
+            elif quant_method == QuantizationMethod.GPTQ:
+                config_class = GPTQConfig
+            elif quant_method == QuantizationMethod.AWQ:
+                config_class = AwqConfig
+            else:
+                # unknown quantization_method, which is not None:
+                raise NotImplementedError(
+                    f"Unsupported quantization method detected: {quant_method}. Check for updates."
+                )
```
what's strange to me is that you have an exact mapping from config to method, the method is just a string basically. I think we should follow what we have in the AutoConfig for example, where you just have `model_type` (here will be the `quant_type` or whatever) and just have a mapping from the string type to the class. 

There's a lot of complexity IMO for something pretty simple + we already commonly use AutoConfigs quite often. The `QuantizationConfig` should inherit form `PushToHubMixin` and have the necessary logic. (be a dataclass we should not need to parse). 
Let's try to re-use what we are familiar with! `GenerationConfig` should help.

ArthurZucker replied:
Then each config will have a type, mapping type to the quantizer kind of like models.

poedator replied:
I agree that it sould be simplified, just have not figured the best way yet. Will try with your hints.

Not sure about inheriting from `PushToHubMixin` - the HFQuantizer will not wrap the model (as Peft does, but will be used as an external tool to process the model).

poedator replied:
refactored this part. Now checking `quantization_config` type instead.

ArthurZucker commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,869 @@
+    return module, tensor_name
+
+
+class HFQuantizer(ABC):
```
If yes, then we should make sure it follows our usual logic: 
1. A quantizer is kind of like a model. We need a better name than HFQuantizer. 
2. We could just load a HFQuantiizer from pretrained, kind of what we do for a model we get the config and init it, we get the quantization method and init whatever elle you need. 
3.

poedator replied:
I viewed HFQuantizer as a tool, and not a model. Something like a `peft` thing.  It may exist separately from models and potentially get passed as an argument in `.from_pretrained()`. So a user would take a custom quantizer (subclassed from HFQuantizer) and load a model using it.

I even thought about making a simple example of like "RoundToNearestQuantizer", but my CUDA skills are not strong yet.

Name-wise - I had no other good ideas. What would you suggest? `QuantizationWrapper`?

ArthurZucker reviewed (commented):
First of all, thanks a lot @poedator, that's a huge and very involved PR! 🤗 It already is miles better from the current implementation! Great job 🙌🏻 

- Great job on cleaning the modeling utils file. The design there is pretty good IMO, parse, quantize, pre-process, post-process. 
- Some things are still a tad bit mixed, special dtypes, device maps etc. If we are able to further isolate these by not having to call the quantizer as often it could be even better! 
- All the introduced types  are not super intuitive for me and add an extra layer of complexity, not needed. 
- I like the logic of QuantizationConfig -> QuantizationWrapper that handles the individual logics. This is very transformers like, where each quantizer backend has it's own logic. 🥳 
- IMO We should not need an additional parser. The QuantizationConfig should be the "parser", following what we do with all other config. We should not need an additional enum and enum mapping for the quantization method as it's basicaly the config_type no?  
- the design of the `Quantizer` is super important as it is setting a new API, but will mostly be used internally, let's keep camel based coding 😈 but also think about potential uses outside `PreTrainedModel.from_pretrained` if you have seen community interests / use cases? 
cc @SunMarc if anything is unclear ping me!

poedator commented on tests/quantization/bnb/test_mixed_int8.py:
```diff
@@ -833,4 +832,4 @@ def test_int8_from_pretrained(self):
         output_sequences = model.generate(input_ids=encoded_input["input_ids"].to(0), max_new_tokens=10)
 
-        self.assertEqual(self.tokenizer.decode(output_sequences[0], skip_special_tokens=True), self.EXPECTED_OUTPUT)
+        self.assertIn(self.tokenizer.decode(output_sequences[0], skip_special_tokens=True), self.EXPECTED_OUTPUTS)
```
I observed that the tests outputs vary depending on the GPU architecture used.  I have A100s and the testing system has something else. 
I see the need for uniform approach here (single vs multiple vs explicitly driven by GPU type correct outputs). It is really up to the maintainers. Also this can be changed in a separate PR.

Just let me keep these multiple outputs for now - they are based on the resutls that come from the main branch, and help me focus on the failing tests.

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -72,8 +72,6 @@
     is_accelerate_available,
-    is_auto_awq_available,
-    is_auto_gptq_available,
     is_bitsandbytes_available,
```
`is_bitsandbytes_available ()` is still used in `modeling_utils.py::ModuleUtilsMixin::num_parameters()`
Agree that it has to go away, but need to refactor this part first.

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,869 @@
+
+logger = logging.get_logger(__name__)
+
+DeviceMap = TypeVar("DeviceMap", str, Dict[str, int])
+QuantizationConfig = TypeVar("QuantizationConfig", QuantizationConfigMixin, Dict)
+StateDict = TypeVar("StateDict", bound=Dict[str, Any])
+MaxMemory = TypeVar("MaxMemory", bound=Dict[Union[int, str], Union[int, str]])
+PreTrainedModel = TypeVar("PreTrainedModel", bound=Any)
```
I wanted to have pretty typing strings, but have no strong feelings about this. Let the style stuff be decided between maintainers. 
I can remove it if you confirm that this is clearly redundant and against HF codestyle.

ArthurZucker replied:
Let's remove it yep 👍🏻

poedator replied:
done

poedator commented:
@ArthurZucker, thank you for the kind words and deep comments!

General reply first: I started this PR trying to repack existing quantization-related code to make it easier adding other quantization methods. What you see by today is effectively the same operations as before but packaged into `HFQuantizer` classes. This explains the numerous calls to quantizer - they remain where I found them. Combining calls to quantizers would be nice, but may require more massive refactoring (and understanding of the code history) and I am feeling a bit out of my depth here. More specific refactoring comments from the maintainers do help me moving forward. 

Another reason why I refrained from deeper refactoring was to keep code familiar to the reviewers.
With this context, would it be possible to postpone deeper refactoring to a separate PR? Although it would be ideal to set the class interface properly now.

Now more on some of the topics that you raised:

The QuantizationMethod enum and config classes like BitsAndBytesConfig were pre-existing.  QuantizationMethod can be removed, but I'd still need a way to keep track of chosen q-method. Maybe it is possible to use config class for that.

re "IMO We should not need an additional parser." - I packaged there the original code that selected `QuantizationMethod`. It should be possible to make it a class method instead. But the HF quantizer is already quite heavy. Currently the parser is a factory for quantizers. 

re "about potential uses outside ... community interests / use cases?" - I envisage possibility of people writing/importing custom quantizers as subclasses of HFQuantizer, and using them to quantize models as they like. Also this PR makes it easier to add more quantization methods, like SPQR which I coauthored recently :)
Perhaps it is time to raise it in Reddit/LocalLlama or other similar place where LLM quantization enthusiasts meet.

poedator committed:
removed typevars in quantizers.py

poedator committed:
cleanup post-rebase 16 jan

poedator committed:
QuantizationConfigParser classmethod refactor

poedator commented on src/transformers/utils/quantizers.py:
```diff
@@ -0,0 +1,869 @@
+    def parse_config_from_args(self, quantization_config_from_args=None, **kwargs) -> Dict[str, Any]:
+        """
+        Parses the quantization configuration from arguments provided to `from_pretrained`.
+        This method sets the `quantization_config` and `quantization_method` based on the provided arguments.
```
Thank for the I refactored this part:
- kept the code to handle bnb quantization config from kwargs (to be deprecated) in `.from_pretrained()`
- made the parser into pure classmethod + few staticmethod auxiliary functions
- combined 2 calls to parser into one.

poedator commented on src/transformers/modeling_utils.py:
```diff
@@ -737,20 +733,17 @@ def _load_state_dict_into_meta_model(
-                    model, param_name, param_device, value=param, fp16_statistics=fp16_statistics
-                )
+            quantizer.create_quantized_param(model, param, param_name, param_device, state_dict)
+            # TODO: consider removing used param_parts from state_dict before return
```
I did that by passing `unexpected keys` to `quantizer.create_quantized_param()`.  not messing ith state dict.
this allows to delete method `HFQuantizer.update_mismatched_keys()`. One fewer call to quantizer.

poedator replied:
This would be even prettier if bnb auxiliary elements were buffers. But it was not the case for 8bit and I also copied that approach to 4bit. Perhaps this will get refactored some day on bnb side.

poedator replied:
I wonder, why `unexpected_keys` becomes a list? it would be more efficient to keep it as a set. @SunMarc

poedator committed:
rework of handling of unexpected aux elements of bnb weights

poedator committed:
moved q-related stuff from save_pretrained to quantizers

younesbelkada committed:
refactor v1

younesbelkada committed:
more changes

younesbelkada committed:
fix some tests

younesbelkada committed:
remove it from main init

younesbelkada committed:
ooops

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

younesbelkada committed:
fix awq issues

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into hf-quantizer-work

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
Merge branch 'main' into hf-quantizer-work

younesbelkada committed:
add docs

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update docs/source/en/hf_quantizer.md

younesbelkada committed:
address comments

younesbelkada committed:
fix

younesbelkada committed:
Merge branch 'hf-quantizer-work' of https://github.com/younesbelkada/transformers into hf-quantizer-work

younesbelkada committed:
fixup

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/modeling_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
address final comment

younesbelkada committed:
Merge branch 'hf-quantizer-work' of https://github.com/younesbelkada/transformers into hf-quantizer-work

younesbelkada committed:
update

younesbelkada committed:
Update src/transformers/quantizers/base.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/quantizers/auto.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
fix

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into hf-quantizer-work

younesbelkada committed:
add kwargs update

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into HEAD

younesbelkada committed:
Merge branch 'quant' into hf-quantizer-work

younesbelkada committed:
fixup

younesbelkada committed:
add `optimum_quantizer` attribute

younesbelkada committed:
oops

younesbelkada committed:
Merge pull request #5 from younesbelkada/hf-quantizer-work

Hf quantizer work

younesbelkada committed:
rm unneeded file

younesbelkada committed:
Merge branch 'quant' of https://github.com/poedator/transformers into HEAD

younesbelkada commented:
cc @ArthurZucker I just ported out the commits from https://github.com/huggingface/transformers/pull/28703 here

younesbelkada reviewed (approved):
Thanks @poedator for your great work on this - the community will largely benefit from this refactor and adding new quantization methods natively in transformers will be much more easy and scalable thanks to your great efforts ! Thanks again for all your time, contribution and patience for this PR and I look forward to collaborating more with you 
I can confirm all quantization slow tests pass with these changes
Thanks also @ArthurZucker for your patience and reviewing https://github.com/huggingface/transformers/pull/28703 ! 🚀

ArthurZucker reviewed (approved):
🚀 Thanks both for the hard work! Looks really nice. 
Let's keep an eye on the tests cc @amyeroberts and @ydshieh

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into HEAD

younesbelkada committed:
fix doctests

ArthurZucker commented:
Merging!

poedator commented:
Thank you, @ArthurZucker  @younesbelkada @SunMarc @amyeroberts and the rest of HF team involved, for getting this to the merge! I was very impressed watching you work on this PR in the past month!

winglian commented on src/transformers/modeling_utils.py:
```diff
@@ -4098,12 +3726,15 @@ def _fix_key(key):
-                    if not (is_quantized):
-                        set_module_tensor_to_device(model, key, "cpu", torch.empty(*param.size(), dtype=target_dtype))
+                    value = torch.empty(*param.size(), dtype=target_dtype)
+                    if getattr(
```
Missed a guard clause here to check `if hf_quantizer is not None and getattr(`

younesbelkada replied:
Thanks ! This should be fixed here: https://github.com/huggingface/transformers/pull/28804 no?
